lucene/core/src/java/org/apache/lucene/index/DefaultIndexingChain.java

```diff
@@ -94,29 +89,100 @@ public DefaultIndexingChain(DocumentsWriterPerThread docWriter) {
+        if (pf == null) {
+          return null;
+        }
+        if (pf.fieldInfo.getDocValuesType() == DocValuesType.SORTED) {
+          return (SortedDocValues) pf.docValuesWriter.getDocValues();
+        }
+        return null;
+      }
+
+      @Override
+      public SortedNumericDocValues getSortedNumericDocValues(String field) throws IOException {
+        PerField pf = getPerField(field);
+        if (pf == null) {
+          return null;
+        }
+        if (pf.fieldInfo.getDocValuesType() == DocValuesType.SORTED_NUMERIC) {
+          return (SortedNumericDocValues) pf.docValuesWriter.getDocValues();
+        }
+        return null;
+      }
+
+      @Override
+      public SortedSetDocValues getSortedSetDocValues(String field) throws IOException {
+        PerField pf = getPerField(field);
+        if (pf == null) {
+          return null;
+        }
+        if (pf.fieldInfo.getDocValuesType() == DocValuesType.SORTED_SET) {
+          return (SortedSetDocValues) pf.docValuesWriter.getDocValues();
+        }
+        return null;
+      }
+
+      @Override
+      public FieldInfos getFieldInfos() {
+        return fieldInfos.finish();
+      }
+
+      @Override
+      public int maxDoc() {
+        return maxDoc;
+      }
+    };
+  }
+
   private Sorter.DocMap maybeSortSegment(SegmentWriteState state) throws IOException {
     Sort indexSort = state.segmentInfo.getIndexSort();
     if (indexSort == null) {
       return null;
     }
 
-    List<Sorter.DocComparator> comparators = new ArrayList<>();
+    LeafReader docValuesReader = getDocValuesReader(state.segmentInfo.maxDoc());
+
+    List<IndexSorter.DocComparator> comparators = new ArrayList<>();
     for (int i = 0; i < indexSort.getSort().length; i++) {
       SortField sortField = indexSort.getSort()[i];
-      PerField perField = getPerField(sortField.getField());
-      if (perField != null && perField.docValuesWriter != null &&
-          finishedDocValues.contains(perField.fieldInfo.name) == false) {
-          perField.docValuesWriter.finish(state.segmentInfo.maxDoc());
-          Sorter.DocComparator cmp = perField.docValuesWriter.getDocComparator(state.segmentInfo.maxDoc(), sortField);
-          comparators.add(cmp);
-          finishedDocValues.add(perField.fieldInfo.name);
-      } else {
-        // safe to ignore, sort field with no values or already seen before
+      IndexSorter sorter = sortField.getIndexSorter();
+      if (sorter == null) {
+        throw new UnsupportedOperationException("Cannot sort index using sort field " + sortField);
       }
+      comparators.add(sorter.getDocComparator(docValuesReader));
     }
     Sorter sorter = new Sorter(indexSort);
     // returns null if the documents are already sorted
-    return sorter.sort(state.segmentInfo.maxDoc(), comparators.toArray(new Sorter.DocComparator[comparators.size()]));
+    return sorter.sort(state.segmentInfo.maxDoc(), comparators.toArray(new IndexSorter.DocComparator[0]));
```

Comment:
There is an even nicer syntax now.

```suggestion
    return sorter.sort(state.segmentInfo.maxDoc(), comparators.toArray(IndexSorter.DocComparator[]::new));
```